### PR TITLE
rfcs: reordering and parallelization of multi-row mutations

### PR DESCRIPTION
**Pull request by knz**
Release note: none

### CODE REVIEW

File: docs/RFCS/order_and_parallelization_of_mutations.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+tl;dr: the data source of a mutation can always be distributed. UPDATE
+and DELETE can be distributed too if there are no FK cascading
+actions. In the other cases, "it depends". The RFC below details these
+cases.
+
+# Motivation
+
+Optimization of queries that include mutation statements over multiple rows.
+
```
Comment (vivekmenezes) [inline docs/RFCS/order_and_parallelization_of_mutations.md]: I appreciate you writing this up, my biggest concern is adding in functionality that users have not asked for. While the project itself is interesting, it adds a lot of complexity to our implementation. Before we consider the implementation details is there enough consensus that we want to do this?

Comment (knz) [reply docs/RFCS/order_and_parallelization_of_mutations.md]: This is not a RFC that drives a change in CockroachDb. This is a RFC that specifies constraints on any future work to do these changes. It's similar in scope to the RFC `20161215_join_filters.md` which specified how the work could be done and was merged long before the change was actually done.

**Comment by knz** [comment]
@andy-kimball @RaduBerinde do you think there is still anything salvageable in this document? When/if we take this information over elsewhere, I can close this PR.

**Comment by RaduBerinde** [comment]
We definitely should have this written down somewhere. Not sure if we have some of this info written down in the opt code somewhere? (CC @andy-kimball).

I don't really understand the statement: "Note that despite its name, the UPDATE statement generally does not need read-then-write, because it is the data source that is in charge of extracting the current values for all the table columns". There is a related example `UPDATE kv SET v = v + 1`; how would you know the value of `v` if you don't read first?

FWIW this is the plan for that example:
```
root@127.20.114.143:36406/defaultdb> explain (opt) update kv set v=v+1 where true;
          text           
+-----------------------+
  update kv              
   └── project           
        ├── scan kv      
        └── projections  
             └── v + 1   
```

**Comment by andy-kimball** [comment]
I don't think this subject is covered anywhere in the opt code. This seems more of a DistSQL planning thing.

**Comment by knz** [comment]
@jordanlewis @RaduBerinde I wonder if this PR has outlived it utility. Is there anything in here that's worth keeping ahead of distributing mutations?

**Comment by RaduBerinde** [comment]
Things have changed drastically since, and they are still in the process of significant changes (e.g. the distsql exec factory). I think we will need to rethink this from scratch when/if we want to do it.

**Comment by knz** [comment]
cool thanks

**Pull request closed without merging**